src/Models/HydrostaticFreeSurfaceModels/implicit_free_surface.jl

```diff
@@ -1 +1,80 @@
+
+### Adapt.adapt_structure(to, free_surface::ExplicitFreeSurface) =
+###     ExplicitFreeSurface(Adapt.adapt(to, free_surface.η), free_surface.gravitational_acceleration)
+
+# Internal function for HydrostaticFreeSurfaceModel
+function FreeSurface(free_surface::ImplicitFreeSurface{Nothing}, velocities, arch, grid)
+    η = CenterField(arch, grid, TracerBoundaryConditions(grid))
+    g = convert(eltype(grid), free_surface.gravitational_acceleration)
+
+    barotropic_x_volume_flux = ReducedField(Face, Center, Nothing, arch, grid; dims=(3), boundary_conditions=nothing)
+    barotropic_y_volume_flux = ReducedField(Center, Face, Nothing, arch, grid; dims=(3), boundary_conditions=nothing)
+    barotropic_volume_flux = (u=barotropic_x_volume_flux, v=barotropic_y_volume_flux)
+
+    Ax_zintegral = ReducedField(Face, Center, Nothing, arch, grid; dims=(3), boundary_conditions=nothing)
+    Ay_zintegral = ReducedField(Center, Face, Nothing, arch, grid; dims=(3), boundary_conditions=nothing)
+    vertically_integrated_lateral_face_areas = (Ax = Ax_zintegral, Ay=Ay_zintegral)
+    compute_vertically_integrated_lateral_face_areas!(vertically_integrated_lateral_face_areas, grid, arch)
+
+    ## implicit_step_solver = ImplicitFreeSurfaceSolver(arch, η, vertically_integrated_lateral_face_areas; maxit=100)
+    implicit_step_solver = ImplicitFreeSurfaceSolver(arch, η, vertically_integrated_lateral_face_areas)
+
+    return ImplicitFreeSurface(η, g, 
+                               barotropic_volume_flux, 
+                               vertically_integrated_lateral_face_areas,
+                               implicit_step_solver)
+end
+
+### In final form the two functions below will return 0 (these are invoked in hydrostatic_free_surface_tendency_kernel_functions.jl )
+explicit_barotropic_pressure_x_gradient(i, j, k, grid, free_surface::ImplicitFreeSurface) =
+    free_surface.gravitational_acceleration * ∂xᶠᵃᵃ(i, j, k, grid, free_surface.η) * 0.0000
+
+explicit_barotropic_pressure_y_gradient(i, j, k, grid, free_surface::ImplicitFreeSurface) =
+    free_surface.gravitational_acceleration * ∂yᵃᶠᵃ(i, j, k, grid, free_surface.η) * 0.0000
+
+@kernel function _compute_vertically_integrated_lateral_face_areas!(grid, A )
+    i, j = @index(Global, NTuple)
+    # U.w[i, j, 1] = 0 is enforced via halo regions.
+    A.Ax[i, j, 1] = 0.
+    A.Ay[i, j, 1] = 0.
+    @unroll for k in 1:grid.Nz
+        @inbounds A.Ax[i, j, 1] += Δyᶠᶜᵃ(i, j, k, grid)*Δzᵃᵃᶜ(i, j, k, grid)
+        @inbounds A.Ay[i, j, 1] += Δxᶜᶠᵃ(i, j, k, grid)*Δzᵃᵃᶜ(i, j, k, grid)
+    end
+end
+
+function compute_vertically_integrated_lateral_face_areas!(vertically_integrated_lateral_face_areas, grid, arch)
+
+    event = launch!(arch,
+                    grid,
+                    :xy,
+                    _compute_vertically_integrated_lateral_face_areas!,
+                    grid,
+                    vertically_integrated_lateral_face_areas,
+                    dependencies=Event(device(arch)))
```

Comment:
The implicit free-surface solver test is failing on GPU and it seems like the error is coming from this call/kernel.

Build log: https://buildkite.com/clima/oceananigans/builds/1506#1c304566-3664-49e7-b56c-e190fbc35183/37-350

It seems to fail in adapting the `ReducedField` for GPU kernels and eventually fails in `validate_reduced_dims(::Nothing)` because `MethodError: no method matching length(::Nothing)`.

Looking at the adapt code: https://github.com/CliMA/Oceananigans.jl/blob/master/src/Fields/reduced_field.jl#L139-L140

It looks like the failure might be due to setting `dims=nothing` when adapting.

Reply:
I'll try fixing this later today to see if the GPU test passes.

Reply:
Ah, one solution to this is to add a new inner constructor that does not validate dims for use with adapt_structure.